assignment-0/controller.go

```diff
@@ -0,0 +1,113 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type controller struct {
+	Client   *kubernetes.Clientset
+	Informer cache.SharedIndexInformer
+}
+
+func newController(client *kubernetes.Clientset, informerFactory informers.SharedInformerFactory) *controller {
+	deploymentInformer := informerFactory.Apps().V1().Deployments().Informer()
+	deploymentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			createService(client, obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			deleteService(client, obj)
+		},
+	})
+
+	return &controller{
+		Client:   client,
+		Informer: deploymentInformer,
+	}
+}
+
+func (c *controller) run(ch <-chan struct{}) {
+	fmt.Println("starting controller")
+	if !cache.WaitForCacheSync(ch, c.Informer.HasSynced) {
+		fmt.Print("waiting for cache to be synced\n")
+	}
+
+	go wait.Until(c.worker, 1*time.Second, ch)
+
+	<-ch
+}
+
+func (c *controller) worker() {
+	for c.processItem() {
+	}
+}
+
+func (c *controller) processItem() bool {
+	return true
+}
+
+func createService(clientset *kubernetes.Clientset, obj interface{}) {
+	deployment := obj.(*appsv1.Deployment)
+	// check if service is present, if not then only create
+	_, err := clientset.CoreV1().Services(deployment.Namespace).Get(
+		context.Background(), deployment.Name+"-service", metav1.GetOptions{})
+	if err == nil {
+		return
+	}
+	fmt.Println("service not present, creating...")
+	// create list of type corev1.ServicePort and assign it to Ports
+	servicePorts := []corev1.ServicePort{}
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		for _, containerPort := range container.Ports {
+			port := corev1.ServicePort{
+				Name:     containerPort.Name,
+				Protocol: containerPort.Protocol,
+				Port:     containerPort.ContainerPort,
+			}
+			servicePorts = append(servicePorts, port)
+		}
+	}
+	_, err = clientset.CoreV1().Services(deployment.Namespace).Create(
+		context.Background(),
+		&corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      deployment.Name + "-service",
+				Namespace: deployment.Namespace,
+				Labels:    deployment.Labels,
+			},
+			Spec: corev1.ServiceSpec{
+				Ports:    servicePorts,
+				Selector: deployment.Labels,
+				Type:     "LoadBalancer",
+			},
+		}, metav1.CreateOptions{})
+	if err != nil {
+		panic(err.Error())
+	}
+}
+
+func deleteService(clientset *kubernetes.Clientset, obj interface{}) {
+	deployment := obj.(*appsv1.Deployment)
+	// check if service is present, if it is then only delete
+	_, err := clientset.CoreV1().Services(deployment.Namespace).Get(
```

Comment:
In this case the entire codebase is in single file, so things are easier but if we had multiple files and we changed how we generate the service name, how would we know that we have to change this place as well. That is the reason we should have a common function to generate the service name and that can be used at both the places.
These things don't seem important but they are, very.

Reply:
fixed